lib/src/layer/polyline_layer.dart

```diff
@@ -127,95 +115,149 @@ class PolylinePainter extends CustomPainter {
+      } else {
+        strokeWidth = polyline.strokeWidth;
+      }
+
+      final isDotted = polyline.isDotted;
+      paint = Paint()
+        ..strokeWidth = strokeWidth
+        ..strokeCap = polyline.strokeCap
+        ..strokeJoin = polyline.strokeJoin
+        ..style = isDotted ? PaintingStyle.fill : PaintingStyle.stroke
+        ..blendMode = BlendMode.srcOver;
+
+      if (polyline.gradientColors == null) {
+        paint.color = polyline.color;
+      } else {
+        polyline.gradientColors!.isNotEmpty
+            ? paint.shader = _paintGradient(polyline, offsets)
+            : paint.color = polyline.color;
+      }
+
+      if (polyline.borderColor != null) {
+        filterPaint = Paint()
+          ..color = polyline.borderColor!.withAlpha(255)
+          ..strokeWidth = strokeWidth
+          ..strokeCap = polyline.strokeCap
+          ..strokeJoin = polyline.strokeJoin
+          ..style = isDotted ? PaintingStyle.fill : PaintingStyle.stroke
+          ..blendMode = BlendMode.dstOut;
+      }
+
+      if (polyline.borderStrokeWidth > 0.0) {
+        borderPaint = Paint()
+          ..color = polyline.borderColor ?? const Color(0x00000000)
+          ..strokeWidth = strokeWidth + polyline.borderStrokeWidth
+          ..strokeCap = polyline.strokeCap
+          ..strokeJoin = polyline.strokeJoin
+          ..style = isDotted ? PaintingStyle.fill : PaintingStyle.stroke
+          ..blendMode = BlendMode.srcOver;
+      }
+
+      final radius = paint.strokeWidth / 2;
+      final borderRadius = (borderPaint?.strokeWidth ?? 0) / 2;
+
       if (saveLayers) canvas.saveLayer(rect, Paint());
-      if (borderPaint != null && filterPaint != null) {
-        borderPaint.style = PaintingStyle.stroke;
-        _paintLine(canvas, polylineOpt.offsets, borderPaint);
-        filterPaint.style = PaintingStyle.stroke;
-        _paintLine(canvas, polylineOpt.offsets, filterPaint);
+      if (isDotted) {
+        final spacing = strokeWidth * 1.5;
+        if (borderPaint != null && filterPaint != null) {
+          _paintDottedLine(borderPath, offsets, borderRadius, spacing);
+          _paintDottedLine(filterPath, offsets, radius, spacing);
+        }
+        _paintDottedLine(path, offsets, radius, spacing);
+      } else {
+        if (borderPaint != null && filterPaint != null) {
+          _paintLine(borderPath, offsets);
+          _paintLine(filterPath, offsets);
+        }
+        _paintLine(path, offsets);
       }
-      _paintLine(canvas, polylineOpt.offsets, paint);
       if (saveLayers) canvas.restore();
     }
+
+    drawPaths();
   }
 
-  void _paintDottedLine(Canvas canvas, List<Offset> offsets, double radius,
-      double stepLength, Paint paint) {
-    final path = ui.Path();
+  void _paintDottedLine(
+      ui.Path path, List<Offset> offsets, double radius, double stepLength) {
     var startDistance = 0.0;
     for (var i = 0; i < offsets.length - 1; i++) {
       final o0 = offsets[i];
```

Comment:
**FOR LINE 265**

For the same performance and code quality reasons as before.

\-      final totalDistance = _dist(o0, o1);
\+      final totalDistance = (o0 - o1).distance;

Using the built in function will let us remove _dist(v, w) and _sqr (x)
Though I cannot suggest removing them in review you will have to create another commit to do that.

Reply:
Done